bootstrap/pkg/kfapp/gcp/gcp.go

```diff
@@ -583,16 +583,209 @@ func setNameVal(entries []configtypes.NameValue, name string, val string, requir
 	return entries
 }
 
-//TODO(#2515)
-func (gcp *Gcp) replaceText(regex string, repl string, src []byte) []byte {
-	re := regexp.MustCompile(regex)
-	buf := re.ReplaceAll(src, []byte(repl))
-	return buf
+// Helper function to generate account field for IAP.
+func (gcp *Gcp) getIapAccount() string {
+	iapAcct := "serviceAccount:" + gcp.Spec.Email
+	if !strings.Contains(gcp.Spec.Email, "iam.gserviceaccount.com") {
+		iapAcct = "user:" + gcp.Spec.Email
+	}
+	return iapAcct
+}
+
+// Write IAM binding rules based on GCP app config.
+func (gcp *Gcp) writeIamBindingsFile(src string, dest string) error {
+	buf, err := ioutil.ReadFile(src)
+	if err != nil {
+		return &kfapis.KfError{
+			Code:    int(kfapis.INTERNAL_ERROR),
+			Message: fmt.Sprintf("Error when reading template %v: %v", src, err),
+		}
+	}
+
+	var data map[string]interface{}
+	if err = yaml.Unmarshal(buf, &data); err != nil {
+		return &kfapis.KfError{
+			Code:    int(kfapis.INTERNAL_ERROR),
+			Message: fmt.Sprintf("Error when unmarshaling template %v: %v", src, err),
+		}
+	}
+
+	e, ok := data["bindings"]
+	if !ok {
+		return &kfapis.KfError{
+			Code:    int(kfapis.INTERNAL_ERROR),
+			Message: "Invalid IAM bindings format: not able to find `bindings` entry.",
+		}
+	}
+
+	roles := map[string]string{
+		"set-kubeflow-admin-service-account": "serviceAccount:" + getSA(gcp.Name, "admin", gcp.Spec.Project),
+		"set-kubeflow-user-service-account":  "serviceAccount:" + getSA(gcp.Name, "user", gcp.Spec.Project),
+		"set-kubeflow-vm-service-account":    "serviceAccount:" + getSA(gcp.Name, "vm", gcp.Spec.Project),
+		"set-kubeflow-iap-account":           gcp.getIapAccount(),
+	}
+
+	bindings := e.([]interface{})
+	for idx, b := range bindings {
+		binding := b.(map[string]interface{})
+		if mem, ok := binding["members"]; ok {
+			members := mem.([]interface{})
+			var newMembers []string
+			for _, m := range members {
+				member := m.(string)
+				if acct, ok := roles[member]; ok {
+					newMembers = append(newMembers, acct)
+				} else {
+					newMembers = append(newMembers, member)
+				}
+			}
+			binding["members"] = newMembers
+			bindings[idx] = binding
+		} else {
+			return &kfapis.KfError{
+				Code:    int(kfapis.INTERNAL_ERROR),
+				Message: "Invalid IAM bindings format: not able to find `members` entry.",
+			}
+		}
+	}
+	data["bindings"] = bindings
+
+	if buf, err = yaml.Marshal(data); err != nil {
+		return &kfapis.KfError{
+			Code:    int(kfapis.INTERNAL_ERROR),
+			Message: fmt.Sprintf("Error when marshaling IAM bindings: %v", err),
+		}
+	}
+	if err = ioutil.WriteFile(dest, buf, 0644); err != nil {
+		return &kfapis.KfError{
+			Code:    int(kfapis.INTERNAL_ERROR),
+			Message: fmt.Sprintf("Error when writing IAM bindings: %v", err),
+		}
+	}
+	return nil
+}
+
+// Replace placeholders and write to cluster-kubeflow.yaml
+func (gcp *Gcp) writeClusterConfig(src string, dest string) error {
```

Comment:
I think "write" is not the word, because we are copying and replacing.
But not sure what's the better option...

Reply:
we are actually reading and writing with replaced file content.